server/src/main/java/com/google/fhir/gateway/FhirUtil.java

```diff
@@ -59,6 +61,27 @@ public static Bundle parseResponseToBundle(FhirContext fhirContext, HttpResponse
     return (Bundle) resource;
   }
 
+  public static IBaseResource createResourceFromRequest(
+      FhirContext fhirContext, RequestDetailsReader request) {
+    byte[] requestContentBytes = request.loadRequestContents();
+    Charset charset = request.getCharset();
+    if (charset == null) {
+      charset = StandardCharsets.UTF_8;
+    }
+    String requestContent = new String(requestContentBytes, charset);
+    IParser jsonParser = fhirContext.newJsonParser();
+    return jsonParser.parseResource(requestContent);
+  }
+
+  public static Bundle parseRequestToBundle(FhirContext fhirContext, RequestDetailsReader request) {
+    IBaseResource resource = createResourceFromRequest(fhirContext, request);
+    if (!(resource instanceof Bundle)) {
```

Comment:
This can be replaced by 
Preconditions.checkArgument(
        FhirUtil.isSameResourceType(resource.fhirType(), ResourceType.Bundle));